Standardization of packages structure and install

tlrx opened the pull request:
The existing DEB/RPM packages have a lot of differences: they don't execute the same actions when installing or removing the package. They also don't declare exactly the same environment variables at the same place. At the end of the day the global behavior and configuration have _few variations_ and it's difficult to maintain the scripts.

This commits unifies the package behavior:
- DEB/RPM use the same package scripts (pre installation, post installation etc) in order to execute exactly the same actions
- Use of a unique environment vars file that declares everything needed by scripts (~~the goal is to delete vars declaration in init.d and systemd scripts, this will be done in another PR~~ see comment below)
- Variables like directory paths are centralized and replaced according to the target platform (using #10330)
- Move /etc/rc.d/init.d to standard /etc/init.d (RPM only)
- Add PID_DIR env var
- Always set ES_USER, ES_GROUP,MAX_MAP_COUNT and MAX_OPEN_FILES in env vars file
- Create log, data, work and plugins directories with DEB/RPM packaging system
- Change to elastic.co domain in copyright and control files
- Add Bats files to automate testing of DEB and RPM packages
- Update TESTING.asciidoc

More info on Bats here:  https://github.com/sstephenson/bats

tlrx committed:
Standardization of packages structure and install

The existing DEB/RPM packages have a lot of differences: they don't execute the same actions when installing or removing the package. They also don't declare exactly the same environment variables at the same place. At the end of the day the global behavior and configuration is *almost* the same but it's very difficult to maintain the scripts.

This commits unifies the package behavior:
- DEB/RPM use the same package scripts (pre installation, post installation etc) in order to execute exactly the same actions
- Use of a unique environment vars file that declares everything needed by scripts (the goal is to delete vars declaration in init.d and systemd scripts, this will be done in another PR)
- Variables like directory paths are centralized and replaced according to the target platform (using #10330)
- Move /etc/rc.d/init.d to standard /etc/init.d (RPM only)
- Add PID_DIR env var
- Always set ES_USER, ES_GROUP,MAX_MAP_COUNT and MAX_OPEN_FILES in env vars file
- Create log, data, work and plugins directories with DEB/RPM packaging system
- Change to elastic.co domain in copyright and control files
- Add Bats files to automate testing of DEB and RPM packages
- Update TESTING.asciidoc

More info on Bats here:  https://github.com/sstephenson/bats

tlrx commented:
@spinscale @electrical Can you have a look please? Thanks :)

tlrx commented:
Note: the Bats test files have been executed with success on Debian 7.8, Debian 8, Ubuntu 12.04, Ubuntu 14.04, OpenSUSE 13, Centos 6.6 and Ferdora 21.

electrical commented:
> the goal is to delete vars declaration in init.d and systemd scripts, this will be done in another PR

This is anti pattern. the init scripts should be able to work even if the defaults/sysconfig files don't exist or are empty.
The defaults/sysconfig file should only function to override defaults, not provide the defaults.

tlrx commented:
@electrical thanks for your precision! I'm OK with self sufficient scripts since #10330 helps to centralize default values in one place.I'd like to see variable definitions placed in specific location(s) and not spread all over the different scripts like it is today :)

To follow your comment maybe I can update this pull request to provide an empty environment file with all known variables names (no default values) each with a comment. Something like this:
```
################################
# Elasticsearch
################################

# Elasticsearch log directory
#LOG_DIR=/var/log/elasticsearch

# Elasticsearch data directory
#DATA_DIR=/var/lib/elasticsearch
...
```

Then in init.d and bin/elasticsearch and bin/plugin scripts we just check if a variable is already known and if not sets a default value if needed.

electrical commented:
As discussed on AoN with @tlrx. 
The env / defaults files should have all possible settings listed but commented.
These files are only used when overriding any default values which are specified in the init / systemd scripts.
It makes sense to ensure we use the same variable setting names across all packages.

Beyond that PR looks good.

tlrx committed:
Update after @eletrical review

tlrx commented:
@electrical I just updated with your comment. I commented all vars in default env vars file and added vars definitions in appropriate scripts. Tests are OK.

electrical commented:
@tlrx looks good as far as i can see.
Would you be able to generate a rpm and deb package from the PR so i can validate those as well?

Thanks!

electrical commented:
All good from my side!

tlrx commented:
@electrical Thanks for your review.

@spinscale Can you have a look please? Thanks :)

spinscale commented on pom.xml:
```diff
@@ -48,12 +48,15 @@
         <packaging.elasticsearch.user>elasticsearch</packaging.elasticsearch.user>
         <packaging.elasticsearch.group>elasticsearch</packaging.elasticsearch.group>
-        <packaging.elasticsearch.work.dir>/var/run/elasticsearch</packaging.elasticsearch.work.dir>
+        <packaging.elasticsearch.work.dir>/tmp/elasticsearch</packaging.elasticsearch.work.dir>
```
What happens if another has created this directory on startup?

tlrx replied:
It looks like `WORK_DIR` is unused by elasticsearch. If another has created this directory, it does not prevent elasticsearch to be installed or to start.

#10672 removes this working directory.

spinscale commented:
tested on debian, installing plugins work, tested on centos, installing plugins work as well

Please make sure we put the bats tests in CI

one problem occured: when installing a plugin that has own configuration files, those are put into `/usr/share/elasticsearch/config` instead of `/etc/elasticsearch`

electrical commented:
@spinscale the issue of plugin installs with configs is caused by the plugin tool not setting the right jvm options.
That's something that should be solved in a separate PR after this is all done.

tlrx commented:
@spinscale @electrical right, it is part of #7946

@spinscale Do you want this issue solved here? And do you want me to add Bats tests that install/remove plugins?

spinscale commented:
@tlrx no need to solve it here, but lets try to fix it soon

LGTM otherwise, CI is next step. Thank you!

tlrx commented:
Merged in 867955188e6e5f86331cbd7ea7273f637ced1f17 and 0dad33f17f3a3dc62e8226cdeced5e17b6976459
